In [5]:
import pandas as pd # A general purpose Python library for data analysis
import numpy as np # A library for scientific computing in Python (e.g., provides high-performance multi-dimensional array objects and operations)

import matplotlib.pyplot as plt # a plotting library for Python and NumPy (readily customizable)
import seaborn as sns # Another plotting library for Python (fewer syntax, excellent default themes, behind the scenes, it uses matplotlib)
import time

## Knowledge Streams 2024

In this notebook, we will learn about the key data structures provided by the Pandas library: **Data Frames, Series, and Indices**.

In addition, we will learn about the following operations:
* How to access data contained in these structures?
* How to read files (e.g., csv, xlsx, sql) to create these structures?
* How to carry out different data manipulation tasks using these structures?

`Dataset`: US elections with information about candidates, their party, votes won, year of election and the result.

## Reading in Data Frames from Files
We'll be using **read_csv** today. Note that this file reading function does all the *data parsing* for you, which is very useful.

Before loading a file into a dataframe, let's first take a look at the **elections.csv** file

In [11]:
#Load csv file and print shape
# Code here
from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv('/content/drive/My Drive/elections.csv')
shape = df.shape
print(shape)

# how many observation and features are given
#Code here
observations = shape[0]
features = shape[1]
print(f"There are {observations} observations and {features} features")

Mounted at /content/drive
(182, 6)
There are 182 observations and 6 features


In [12]:
# We can use the **head command** to show only a few rows of a dataframe from start.
# Code here
print(df.head(10))

#Use **tail command** to show last few observation.
# code here
print(df.tail(10))

   Year               Candidate                  Party  Popular vote Result  \
0  1824          Andrew Jackson  Democratic-Republican        151271   loss   
1  1824       John Quincy Adams  Democratic-Republican        113142    win   
2  1828          Andrew Jackson             Democratic        642806    win   
3  1828       John Quincy Adams    National Republican        500897   loss   
4  1832          Andrew Jackson             Democratic        702735    win   
5  1832              Henry Clay    National Republican        484205   loss   
6  1832            William Wirt           Anti-Masonic        100715   loss   
7  1836       Hugh Lawson White                   Whig        146109   loss   
8  1836        Martin Van Buren             Democratic        763291    win   
9  1836  William Henry Harrison                   Whig        550816   loss   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789  
5  37.603628  
6   7.821583  
7  10.0059

In [13]:
#The `read_csv` command lets us specify a **column to use an index**. For example, we could have used __Year__ as the index.
#Code here
data = pd.read_csv('/content/drive/My Drive/elections.csv', index_col="Year")
print(data)

              Candidate                  Party  Popular vote Result          %
Year                                                                          
1824     Andrew Jackson  Democratic-Republican        151271   loss  57.210122
1824  John Quincy Adams  Democratic-Republican        113142    win  42.789878
1828     Andrew Jackson             Democratic        642806    win  56.203927
1828  John Quincy Adams    National Republican        500897   loss  43.796073
1832     Andrew Jackson             Democratic        702735    win  54.574789
...                 ...                    ...           ...    ...        ...
2016         Jill Stein                  Green       1457226   loss   1.073699
2020       Joseph Biden             Democratic      81268924    win  51.311515
2020       Donald Trump             Republican      74216154   loss  46.858542
2020       Jo Jorgensen            Libertarian       1865724   loss   1.177979
2020     Howard Hawkins                  Green      

In [14]:
#Alternately, we could have used the **set_index** commmand on the dataframe to set a particular column as index.
# code here
data.reset_index(inplace=True)
data.set_index("Year")


Candidate                  Party  Popular vote Result          %
Year                                                                          
1824     Andrew Jackson  Democratic-Republican        151271   loss  57.210122
1824  John Quincy Adams  Democratic-Republican        113142    win  42.789878
1828     Andrew Jackson             Democratic        642806    win  56.203927
1828  John Quincy Adams    National Republican        500897   loss  43.796073
1832     Andrew Jackson             Democratic        702735    win  54.574789
...                 ...                    ...           ...    ...        ...
2016         Jill Stein                  Green       1457226   loss   1.073699
2020       Joseph Biden             Democratic      81268924    win  51.311515
2020       Donald Trump             Republican      74216154   loss  46.858542
2020       Jo Jorgensen            Libertarian       1865724   loss   1.177979
2020     Howard Hawkins                  Green        405035   loss   0.255731

[182 rows x 5 columns]

# Caution:
The **set_index command** (along with all other data frame methods) **does not modify the dataframe**, i.e., the original "elections" is untouched. Note: There is a flag called "inplace" which does modify the calling dataframe (e.g., `elections.set_index("Party",inplace=True)`).

## Duplicate Columns?
By contast, column names MUST be unique. For example, if we try to read in a file for which column names are not unique, Pandas will automatically any duplicates. Load duplicate_columns.csv

In [15]:
#Answer Here
duplicate_file = pd.read_csv('/content/drive/My Drive/duplicate_columns.csv')
print(duplicate_file)

    name    name.1      flavor
0   john     smith     vanilla
1  zhang      shan   chocolate
2  fulan  alfulani  strawberry
3   hong   gildong      banana


## The [ ] Operator & Indexing

The DataFrame class has an indexing operator **[ ]** (also known as the 'brack' operator) that lets you do a variety of different things. If your provide a String to the **[ ]** operator, you get back a ***Series*** corresponding to the requested label.

1.Use **[ ]** to display different columns

2.Use List retrive multiple columns

In [16]:
# Answer Here
column_list1 = data['Candidate'].tolist()
column_list2 = data['Party'].tolist()
print(column_list1)
print(column_list2)

['Andrew Jackson', 'John Quincy Adams', 'Andrew Jackson', 'John Quincy Adams', 'Andrew Jackson', 'Henry Clay', 'William Wirt', 'Hugh Lawson White', 'Martin Van Buren', 'William Henry Harrison', 'Martin Van Buren', 'William Henry Harrison', 'Henry Clay', 'James Polk', 'Lewis Cass', 'Martin Van Buren', 'Zachary Taylor', 'Franklin Pierce', 'John P. Hale', 'Winfield Scott', 'James Buchanan', 'John C. Frémont', 'Millard Fillmore', 'Abraham Lincoln', 'John Bell', 'John C. Breckinridge', 'Stephen A. Douglas', 'Abraham Lincoln', 'George B. McClellan', 'Horatio Seymour', 'Ulysses Grant', 'Horace Greeley', 'Ulysses Grant', 'Rutherford Hayes', 'Samuel J. Tilden', 'James B. Weaver', 'James Garfield', 'Winfield Scott Hancock', 'Benjamin Butler', 'Grover Cleveland', 'James G. Blaine', 'John St. John', 'Alson Streeter', 'Benjamin Harrison', 'Clinton B. Fisk', 'Grover Cleveland', 'Benjamin Harrison', 'Grover Cleveland', 'James B. Weaver', 'John Bidwell', 'John M. Palmer', 'Joshua Levering', 'William J

The **[ ]** operator also accepts a list of strings. In this case, you get back a **DataFrame** corresponding to the requested strings.

In [17]:
# Answer Here
data[['Candidate', 'Party']]

Candidate                  Party
0       Andrew Jackson  Democratic-Republican
1    John Quincy Adams  Democratic-Republican
2       Andrew Jackson             Democratic
3    John Quincy Adams    National Republican
4       Andrew Jackson             Democratic
..                 ...                    ...
177         Jill Stein                  Green
178       Joseph Biden             Democratic
179       Donald Trump             Republican
180       Jo Jorgensen            Libertarian
181     Howard Hawkins                  Green

[182 rows x 2 columns]

A list of one label also returns a DataFrame. This can be handy if you want your results as a DataFrame, not a series.

Note that we can also use the **to_frame** method to turn a Series into a DataFrame.

Extract one col name "Candidates" from DataFrame it will be a series. Convert series into a DataFrame.

In [18]:
# Answer Here
candidate = data['Candidate']
print(candidate.head(10))
candidate.to_frame().head(10)

0            Andrew Jackson
1         John Quincy Adams
2            Andrew Jackson
3         John Quincy Adams
4            Andrew Jackson
5                Henry Clay
6              William Wirt
7         Hugh Lawson White
8          Martin Van Buren
9    William Henry Harrison
Name: Candidate, dtype: object


Candidate
0          Andrew Jackson
1       John Quincy Adams
2          Andrew Jackson
3       John Quincy Adams
4          Andrew Jackson
5              Henry Clay
6            William Wirt
7       Hugh Lawson White
8        Martin Van Buren
9  William Henry Harrison

### Row Indexing

The `[]` operator also accepts numerical slices as arguments. In this case, we are indexing by row, not column!

Extract few rows from DataFrame

In [19]:
# Answer Here
df_subset = df[:5]
print(df_subset)

   Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789  


If you provide a single argument to the `[]` operator, it tries to use it as a name. This is true even if the argument passed to **[ ]** is an integer.

In [20]:
#elections[0] #this does not work, try uncommenting this to see it fail in action, woo
df_subset = df[0]
print(df_subset)

KeyError: 0

The following cells allow you to **test your understanding**. Let's go over the summary of what we have learnt (see slides).

# Creating DataFrames
Create DataFrame using List and Columns name.

In [21]:
# Answer Here
df1 = pd.DataFrame([[1,'one'],[2,'two'],[3,'three']], columns = ['Number', 'Description'])
print(df1)

   Number Description
0       1         one
1       2         two
2       3       three


Creating DataFrames using **Dictionary**.

In [22]:
# Answer Here
df2 = pd.DataFrame({"Fruits":['Strawberry', 'Orange'], "Price":[5.67, 9.65]})
print(df2)

       Fruits  Price
0  Strawberry   5.67
1      Orange   9.65


## Filtering via Boolean Array Selection

The `[]` operator also supports array of booleans as an input. In this case, the array must be exactly as long as the number of rows. The result is a **filtered version of the data frame**, where **only rows corresponding to True appear**.

In [9]:
df1[[True,False,True]]

Number Description
0       1         one
2       3       three

One very common task in Data Science is **filtering**. Boolean Array Selection is one way to achieve this in Pandas. We start by observing that **logical operators** like the equality operator can be applied to **Pandas Series data** to generate a **Boolean Array**.

Compare the 'Result' column to the String 'win' and Show results

In [23]:
#Answer Here
filter_data = data.Result == "win"
data[filter_data].head()

Year               Candidate                  Party  Popular vote Result  \
1   1824       John Quincy Adams  Democratic-Republican        113142    win   
2   1828          Andrew Jackson             Democratic        642806    win   
4   1832          Andrew Jackson             Democratic        702735    win   
8   1836        Martin Van Buren             Democratic        763291    win   
11  1840  William Henry Harrison                   Whig       1275583    win   

            %  
1   42.789878  
2   56.203927  
4   54.574789  
8   52.272472  
11  53.051213

Compare the 'Party' column to the String 'Democratic' and Show results

In [26]:
#Answer Here
filter_party_data = data.Party == "Democratic"
data[filter_party_data].head()

Year         Candidate       Party  Popular vote Result          %
2   1828    Andrew Jackson  Democratic        642806    win  56.203927
4   1832    Andrew Jackson  Democratic        702735    win  54.574789
8   1836  Martin Van Buren  Democratic        763291    win  52.272472
10  1840  Martin Van Buren  Democratic       1128854   loss  46.948787
13  1844        James Polk  Democratic       1339570    win  50.749477

The output of the logical operator applied to the Series is **another Series with the same name and index, but of datatype boolean**.

These boolean Series can be used as an argument to the `[]` operator.

Creates  DataFrame of all election winners since 1980.

In [30]:
#Answer Here
is_win = data[(data.Result == "win") & (data.Year >= 1980)]
is_win

Year          Candidate       Party  Popular vote Result          %
131  1980      Ronald Reagan  Republican      43903230    win  50.897944
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
140  1992       Bill Clinton  Democratic      44909806    win  43.118485
144  1996       Bill Clinton  Democratic      47400125    win  49.296938
152  2000     George W. Bush  Republican      50456002    win  47.974666
157  2004     George W. Bush  Republican      62040610    win  50.771824
162  2008       Barack Obama  Democratic      69498516    win  53.023510
168  2012       Barack Obama  Democratic      65915795    win  51.258484
173  2016       Donald Trump  Republican      62984828    win  46.407862
178  2020       Joseph Biden  Democratic      81268924    win  51.311515

Above, we've assigned the result of the logical operator to a new variable called `iswin`. This is uncommon. Usually, the series is created and used on the same line. Such code is a little tricky to read at first, but you'll get used to it quickly.

Show all 'win' results between 1980 to 2000

In [32]:
#Answer Here
#Answer Here
is_win = data[(data.Result == "win") & (data.Year >= 1980) & (data.Year <= 2000)]
is_win

Year          Candidate       Party  Popular vote Result          %
131  1980      Ronald Reagan  Republican      43903230    win  50.897944
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
140  1992       Bill Clinton  Democratic      44909806    win  43.118485
144  1996       Bill Clinton  Democratic      47400125    win  49.296938
152  2000     George W. Bush  Republican      50456002    win  47.974666

Show all 'Loss' results of Independent party

In [33]:
# Answer Here
loss_results = data[(data.Party == "Independent") & (data.Result == "loss")]
loss_results

Year         Candidate        Party  Popular vote Result          %
121  1976   Eugene McCarthy  Independent        740460   loss   0.911649
130  1980  John B. Anderson  Independent       5719850   loss   6.631143
143  1992        Ross Perot  Independent      19743821   loss  18.956298
161  2004       Ralph Nader  Independent        465151   loss   0.380663
167  2008       Ralph Nader  Independent        739034   loss   0.563842
174  2016     Evan McMullin  Independent        732273   loss   0.539546

We can select multiple criteria by creating multiple boolean Series and combining them using the `&` operator.

Show results of win with percentage less than 50%

In [35]:
# Answer Here
win_results = data.Result == "win"
win_percentage = data["%"] < 50
results = data[win_results & win_percentage]
results.head()

Year          Candidate                  Party  Popular vote Result  \
1   1824  John Quincy Adams  Democratic-Republican        113142    win   
16  1848     Zachary Taylor                   Whig       1360235    win   
20  1856     James Buchanan             Democratic       1835140    win   
23  1860    Abraham Lincoln             Republican       1855993    win   
33  1876   Rutherford Hayes             Republican       4034142    win   

            %  
1   42.789878  
16  47.309296  
20  45.306080  
23  39.699408  
33  48.471624

Show all 'win' results between 1980 to 2000

In [36]:
# Answer Here
res=data[(data.Result=="win") & (data.Year>1980) & (data.Year<2000)]
res

Year          Candidate       Party  Popular vote Result          %
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
140  1992       Bill Clinton  Democratic      44909806    win  43.118485
144  1996       Bill Clinton  Democratic      47400125    win  49.296938

## Loc and iLoc

Show 5 enteries from start

In [42]:
# Answer Here
data.head()
data.loc[:5]

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   
5  1832         Henry Clay    National Republican        484205   loss   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789  
5  37.603628

You can provide `.loc` a list of row labels [0-5] and column labels ['Candidate','Party', 'Year'] as input to return a dataframe

In [53]:
#Answer Here
results = data.loc[[0,1,2,3,4,5], ['Candidate', 'Party', 'Year']]
results

Candidate                  Party  Year
0     Andrew Jackson  Democratic-Republican  1824
1  John Quincy Adams  Democratic-Republican  1824
2     Andrew Jackson             Democratic  1828
3  John Quincy Adams    National Republican  1828
4     Andrew Jackson             Democratic  1832
5         Henry Clay    National Republican  1832

Loc also supports **slicing** (for all types, including numeric and string labels!). Note that the slicing for loc is **inclusive**, even for numeric slices.

Use Slicing on Rows and Columns

In [61]:
# Answer Here
result_1 = data.loc[:5, ['Candidate', 'Party', 'Year']]
result_2 = data.iloc[:6, :3]
result_2

Year          Candidate                  Party
0  1824     Andrew Jackson  Democratic-Republican
1  1824  John Quincy Adams  Democratic-Republican
2  1828     Andrew Jackson             Democratic
3  1828  John Quincy Adams    National Republican
4  1832     Andrew Jackson             Democratic
5  1832         Henry Clay    National Republican

If we provide only a **single label** for the column argument, we get back a **Series**.

In [71]:
# Answer Here
data.loc[:5, 'Candidate']


0       Andrew Jackson
1    John Quincy Adams
2       Andrew Jackson
3    John Quincy Adams
4       Andrew Jackson
5           Henry Clay
Name: Candidate, dtype: object

If we want a data frame instead and don't want to use to_frame, we can provide a **list** containing the column name.

In [74]:
# Answer Here
results = data.loc[:5, ['Candidate']]
results

Candidate
0     Andrew Jackson
1  John Quincy Adams
2     Andrew Jackson
3  John Quincy Adams
4     Andrew Jackson
5         Henry Clay

If we give only one row but many column labels, we'll get back a **Series** corresponding to a row of the table. This new Series has a neat index, where **each entry is the name of the column** that the data came from.

In [76]:
# Answer Here
results = data.loc[0, ['Candidate', 'Party', 'Year']]
results

Candidate           Andrew Jackson
Party        Democratic-Republican
Year                          1824
Name: 0, dtype: object

In [ ]:
# Answer Here

If we omit the column argument altogether, the **default behavior is to retrieve all columns**.

In [78]:
# Answer Here
results = data.loc[0, ]
results

Year                             1824
Candidate              Andrew Jackson
Party           Democratic-Republican
Popular vote                   151271
Result                           loss
%                           57.210122
Name: 0, dtype: object

Specify Rows and Columns as List to retrive specific enteries

In [80]:
# Answer Here
results = data.loc[[0,2,3], ['Candidate', 'Party', 'Year']]
results

Candidate                  Party  Year
0     Andrew Jackson  Democratic-Republican  1824
2     Andrew Jackson             Democratic  1828
3  John Quincy Adams    National Republican  1828

Boolean Series are also boolean arrays, so we can use the Boolean Array Selection from earlier using loc as well.

In [86]:
# Answer Here
data_set = data.Result == "win"
data.loc[data_set]
data.loc[data.Result == "win", ['Candidate', 'Party']].head()


Candidate                  Party
1        John Quincy Adams  Democratic-Republican
2           Andrew Jackson             Democratic
4           Andrew Jackson             Democratic
8         Martin Van Buren             Democratic
11  William Henry Harrison                   Whig

## String-labeled Rows

Let's do a quick example using data with string-labeled rows instead of integer labeled rows, just to make sure we're really understanding loc.

Use mottos.csv file

In [91]:
# Answer Here
df=pd.read_csv('/content/drive/My Drive/mottos.csv',index_col="State")
df.loc[['Alaska','Colorado'],['Motto','Translation']]
df.loc["Alabama":"California","Motto":"Language"]

Motto                 Translation Language
State                                                                         
Alabama     Audemus jura nostra defendere  We dare defend our rights!    Latin
Alaska                North to the future                           —  English
Arizona                        Ditat Deus                God enriches    Latin
Arkansas                   Regnat populus             The people rule    Latin
California                Eureka (Εὕρηκα)             I have found it    Greek

Extract slice, can be specified using slice notation, even if the rows have string labels instead of integer labels.

### iloc

loc's cousin iloc is very similar, but is used to access based on numerical position instead of label. For example, to access to the top 3 rows and top 3 columns of a table, we can use [0:3, 0:3]. 'iloc' slicing is **exclusive**, just like standard Python slicing of numerical values.

Use iloc to extract first 3 rows and columns from elections DataFrame

In [92]:
#Answer Here
df.iloc[0:3, 0:3]

Motto                 Translation Language
State                                                                      
Alabama  Audemus jura nostra defendere  We dare defend our rights!    Latin
Alaska             North to the future                           —  English
Arizona                     Ditat Deus                God enriches    Latin

We will use both `loc` and `iloc` in the course. `loc` is generally preferred for a number of reasons, for example:

1. It is harder to make mistakes since you have to literally write out what you want to get.
2. Code is easier to read, because the reader doesn't have to know e.g., what column #17 represents.
3. It is robust against permutations of the data, e.g. the social security administration switches the order of two columns.

However, iloc is sometimes more convenient. We'll provide examples of when iloc is the superior choice.

## Handy Properties and Utility Functions for Series and DataFrames

The head, shape, size, and describe methods can be used to quickly get a good sense of the data we're working with. For example:

In [93]:
mottos = pd.read_csv("/content/drive/MyDrive/mottos.csv")

In [95]:
# Answer Here
print(mottos.shape)
mottos.head()

(50, 5)


State                          Motto                 Translation  \
0     Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1      Alaska            North to the future                           —   
2     Arizona                     Ditat Deus                God enriches   
3    Arkansas                 Regnat populus             The people rule   
4  California                Eureka (Εὕρηκα)             I have found it   

  Language Date Adopted  
0    Latin         1923  
1  English         1967  
2    Latin         1863  
3    Latin         1907  
4    Greek         1849

Size of DataFrame

In [97]:
# Answer Here
mottos.size

250

The fact that the size is 250 means our data file is relatively small, with only 250 total entries.

Shape of DataFrame

In [98]:
# Answer Here
mottos.shape

(50, 5)

Use describe function and extract the meaningful information from DataFrame

In [99]:
# Answer Here
info=mottos.describe()
info

State                          Motto Translation Language  \
count        50                             50          49       50   
unique       50                             50          30        8   
top     Alabama  Audemus jura nostra defendere           —    Latin   
freq          1                              1          20       23   

       Date Adopted  
count            50  
unique           47  
top            1893  
freq              2

Above, we see a quick summary of all the data. For example, the most common language for mottos is Latin, which covers 23 different states. Does anything else seem surprising?

We can get a direct reference to the index using .index.

In [101]:
# Answer Here
info.index[1]

'unique'

In [102]:
mottos.head(2)

State                          Motto                 Translation  \
0  Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1   Alaska            North to the future                           —   

  Language Date Adopted  
0    Latin         1923  
1  English         1967

It turns out the columns also have an Index. We can access this index by using `.columns`.

In [103]:
# Answer Here
info[:]

State                          Motto Translation Language  \
count        50                             50          49       50   
unique       50                             50          30        8   
top     Alabama  Audemus jura nostra defendere           —    Latin   
freq          1                              1          20       23   

       Date Adopted  
count            50  
unique           47  
top            1893  
freq              2

## Sorting and Value Counts

There are also a ton of useful utility methods we can use with Data Frames and Series. For example, we can create a copy of a data frame sorted by a specific column using `sort_values`.

In [ ]:
# Answer Here

As mentioned before, all Data Frame methods return a copy and do **not** modify the original data structure, unless you set inplace to True.

If we want to sort in reverse order, we can set `ascending=False`.

In [104]:
data.sort_values('%', ascending=False)

Year           Candidate         Party  Popular vote Result          %
114  1964      Lyndon Johnson    Democratic      43127041    win  61.344703
91   1936  Franklin Roosevelt    Democratic      27752648    win  60.978107
120  1972       Richard Nixon    Republican      47168710    win  60.907806
79   1920      Warren Harding    Republican      16144093    win  60.574501
133  1984       Ronald Reagan    Republican      54455472    win  59.023326
..    ...                 ...           ...           ...    ...        ...
165  2008    Cynthia McKinney         Green        161797   loss   0.123442
148  1996        John Hagelin   Natural Law        113670   loss   0.118219
160  2004    Michael Peroutka  Constitution        143630   loss   0.117542
141  1992            Bo Gritz      Populist        106152   loss   0.101918
156  2004          David Cobb         Green        119859   loss   0.098088

[182 rows x 6 columns]

We can also use `sort_values` on Series objects.

In [105]:
mottos['Language'].sort_values().head(50)

46    Chinook Jargon
49           English
29           English
28           English
27           English
26           English
48           English
37           English
38           English
40           English
17           English
34           English
42           English
14           English
41           English
12           English
1            English
13           English
8            English
7            English
9            English
43           English
22            French
4              Greek
10          Hawaiian
19           Italian
39             Latin
44             Latin
36             Latin
45             Latin
47             Latin
35             Latin
33             Latin
0              Latin
31             Latin
30             Latin
23             Latin
21             Latin
20             Latin
18             Latin
16             Latin
15             Latin
11             Latin
6              Latin
5              Latin
3              Latin
2              Latin
32             Latin
24             Latin
25           Spanish
Name: Language, dtype: object

For Series, the `value_counts` method is often quite handy.

In [106]:
mottos['Language'].value_counts()

Language
Latin             23
English           21
Greek              1
Hawaiian           1
Italian            1
French             1
Spanish            1
Chinook Jargon     1
Name: count, dtype: int64

Also commonly used is the `unique` method, which returns **all unique values** as a numpy array.

In [107]:
mottos['Language'].unique()

array(['Latin', 'English', 'Greek', 'Hawaiian', 'Italian', 'French',
       'Spanish', 'Chinook Jargon'], dtype=object)

In [108]:
def fiba(n):
    if n < 2:
        return n
    else:
        return fiba(n-1) + fiba(n-2)



fiba(5)

5

# Thank you!